auth-server flaky test

lib/cad-reminders create process removed second reminder from redis correctly - removed second reminder from redis correctly

https://app.circleci.com/pipelines/github/mozilla/fxa/13129/workflows/2881c192-9ebc-4317-ae9f-c542e5689b38/jobs/189835



┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXA-2485)

➤ Vijay Budhram commented:

Another new one

>     lib/senders/sms: initialise: invalid data: wait a tick: called log.error correctly - called log.error correctly
    /home/circleci/project/packages/fxa-auth-server/test/local/senders/sms.js

AssertionError: expected 2 to equal 1
    at Context.<anonymous> (test/local/senders/sms.js:205:18)
    at processImmediate (internal/timers.js:456:21)https://app.circleci.com/pipelines/github/mozilla/fxa/13932/workflows/28039d32-3089-43d3-9eae-884d1a73136b/jobs/196314

--- a/packages/fxa-auth-server/lib/cad-reminders.js
+++ b/packages/fxa-auth-server/lib/cad-reminders.js
@@ -60,10 +60,9 @@ class CadReminders {
    *                      of elements added to that sorted set, i.e. the result of
    *                      [`redis.zadd`](https://redis.io/commands/zadd).
    */
-  async create(uid) {
+  async create(uid, now = Date.now()) {
     try {
       if (this.rolloutRate <= 1 && Math.random() < this.rolloutRate) {
-        const now = Date.now();
         const result = await props(
           this.keys.reduce((result, key) => {
             result[key] = this.redis.zadd(key, now, uid);
@@ -149,9 +148,8 @@ class CadReminders {
    *                      { timestamp, uid } reminder records
    *                      that have ticked past the relevant expiry interval.
    */
-  async process() {
+  async process(now = Date.now()) {
     try {
-      const now = Date.now();
       return await props(
         this.keys.reduce((result, key) => {
           const cutoff = now - this.intervals[key];

--- a/packages/fxa-auth-server/lib/verification-reminders.js
+++ b/packages/fxa-auth-server/lib/verification-reminders.js
@@ -107,10 +107,9 @@ module.exports = (log, config) => {
      *                      of elements added to that sorted set, i.e. the result of
      *                      [`redis.zadd`](https://redis.io/commands/zadd).
      */
-    async create(uid, flowId, flowBeginTime) {
+    async create(uid, flowId, flowBeginTime, now = Date.now()) {
       try {
         if (rolloutRate <= 1 && Math.random() < rolloutRate) {
-          const now = Date.now();
           const result = await props(
             keys.reduce((result, key) => {
               result[key] = redis.zadd(key, now, uid);
@@ -174,9 +173,8 @@ module.exports = (log, config) => {
      *                      { timestamp, uid, flowId, flowBeginTime } reminder records
      *                      that have ticked past the relevant expiry interval.
      */
-    async process() {
+    async process(now = Date.now()) {
       try {
-        const now = Date.now();
         return await props(
           keys.reduce((result, key, keyIndex) => {
             const cutoff = now - intervals[key];

--- a/packages/fxa-auth-server/package.json
+++ b/packages/fxa-auth-server/package.json
@@ -152,6 +152,8 @@
     "babel-loader": "^8.2.2",
     "binary-split": "1.0.5",
     "chai": "^4.3.4",
+    "esbuild": "^0.12.22",
+    "esbuild-register": "^3.0.0",
     "eslint": "^7.29.0",
     "eslint-plugin-fxa": "^2.0.2",
     "fxa-auth-db-mysql": "workspace:*",

--- a/packages/fxa-auth-server/package.json
+++ b/packages/fxa-auth-server/package.json
@@ -25,7 +25,7 @@
     "format": "prettier --write --config ../../_dev/.prettierrc '**'",
     "gen-keys": "ts-node ./scripts/gen_keys.js; ts-node ./scripts/oauth_gen_keys.js; ts-node ./scripts/gen_vapid_keys.js",
     "write-emails": "npm run emails-scss && node -r ts-node/register ./scripts/write-emails-to-disk.js",
-    "clean-up-old-ci-stripe-customers": "ts-node ./scripts/clean-up-old-ci-stripe-customers.js",
+    "clean-up-old-ci-stripe-customers": "node -r esbuild-register ./scripts/clean-up-old-ci-stripe-customers.js --limit 1000",
     "populate-vat-taxes": "CONFIG_FILES='config/secrets.json' ts-node ./scripts/populate-vat-taxes.ts",
     "paypal-processor": "CONFIG_FILES='config/secrets.json' ts-node ./scripts/paypal-processor.ts",
     "subscription-reminders": "CONFIG_FILES='config/secrets.json' ts-node ./scripts/subscription-reminders.ts",

--- a/packages/fxa-auth-server/scripts/clean-up-old-ci-stripe-customers.js
+++ b/packages/fxa-auth-server/scripts/clean-up-old-ci-stripe-customers.js
@@ -79,7 +79,10 @@ async function init() {
   Container.set(CurrencyHelper, currencyHelper);
 
   const { createStripeHelper } = require('../lib/payments/stripe');
-  const stripeHelper = createStripeHelper(log, config, null);
+  const stripeHelper = createStripeHelper(log, config, {
+    timing: () => null,
+    increment: () => null,
+  });
   const stripe = stripeHelper.stripe;
 
   const listOpts = {};

--- a/packages/fxa-auth-server/scripts/test-ci.sh
+++ b/packages/fxa-auth-server/scripts/test-ci.sh
@@ -8,14 +8,14 @@ yarn workspaces focus fxa-auth-server
 export NODE_ENV=dev
 export CORS_ORIGIN="http://foo,http://bar"
 
-DEFAULT_ARGS="--require ts-node/register --recursive --timeout 5000 --exit --reporter mocha-junit-reporter"
+DEFAULT_ARGS="--require esbuild-register --recursive --timeout 5000 --exit --reporter mocha-junit-reporter"
 
-node -r ts-node/register ./scripts/gen_keys.js
-node -r ts-node/register ./scripts/gen_vapid_keys.js
-node -r ts-node/register ./scripts/oauth_gen_keys.js
+node -r esbuild-register ./scripts/gen_keys.js
+node -r esbuild-register ./scripts/gen_vapid_keys.js
+node -r esbuild-register ./scripts/oauth_gen_keys.js
 ../../_scripts/check-mysql.sh
-node -r ts-node/register ../fxa-auth-db-mysql/bin/db_patcher > /dev/null
-node -r ts-node/register ./scripts/oauth-db-patcher.js
+node -r esbuild-register ../fxa-auth-db-mysql/bin/db_patcher > /dev/null
+node -r esbuild-register ./scripts/oauth-db-patcher.js
 
 TESTS=(local oauth remote scripts)
 for t in "${TESTS[@]}"; do

--- a/packages/fxa-auth-server/scripts/test-local.sh
+++ b/packages/fxa-auth-server/scripts/test-local.sh
@@ -1,6 +1,7 @@
-#!/bin/sh
+#!/bin/bash -e
 
-set -e
+DIR=$(dirname "$0")
+cd "$DIR/.."
 
 rm -rf coverage
 rm -rf .nyc_output
@@ -8,30 +9,38 @@ rm -rf .nyc_output
 if [ -z "$NODE_ENV" ]; then export NODE_ENV=dev; fi;
 if [ -z "$CORS_ORIGIN" ]; then export CORS_ORIGIN="http://foo,http://bar"; fi;
 
-set -u
 
-DEFAULT_ARGS="--require ts-node/register --recursive --timeout 5000 --exit"
+DEFAULT_ARGS="--require esbuild-register --recursive --timeout 5000 --exit"
 
-node -r ts-node/register ./scripts/gen_keys.js
-node -r ts-node/register ./scripts/gen_vapid_keys.js
-node -r ts-node/register ./scripts/oauth_gen_keys.js
-node -r ts-node/register ../fxa-auth-db-mysql/bin/db_patcher > /dev/null
-node -r ts-node/register ./scripts/oauth-db-patcher.js
+if [[ ! -e config/secret-key.json ]]; then
+  node -r esbuild-register ./scripts/gen_keys.js
+fi
+
+if [[ ! -e config/vapid-keys.json ]]; then
+  node -r esbuild-register ./scripts/gen_vapid_keys.js
+fi
+
+if [[ ! -e config/key.json ]]; then
+  node -r esbuild-register ./scripts/oauth_gen_keys.js
+fi
+
+node -r esbuild-register ../fxa-auth-db-mysql/bin/db_patcher > /dev/null
+node -r esbuild-register ./scripts/oauth-db-patcher.js
 
 GLOB=$*
 if [ -z "$GLOB" ]; then
   echo "Local tests"
-  ./scripts/mocha-coverage.js $DEFAULT_ARGS test/local
+  mocha $DEFAULT_ARGS test/local
 
   echo "Oauth tests"
-  ./scripts/mocha-coverage.js $DEFAULT_ARGS test/oauth
+  mocha $DEFAULT_ARGS test/oauth
 
   echo "Remote tests"
-  ./scripts/mocha-coverage.js $DEFAULT_ARGS test/remote
+  mocha $DEFAULT_ARGS test/remote
 
   echo "Script tests"
-  ./scripts/mocha-coverage.js $DEFAULT_ARGS test/scripts
+  mocha $DEFAULT_ARGS test/scripts
 
 else
-  ./scripts/mocha-coverage.js $DEFAULT_ARGS $GLOB
+  mocha $DEFAULT_ARGS $GLOB
 fi

--- a/packages/fxa-auth-server/test/local/cad-reminders.js
+++ b/packages/fxa-auth-server/test/local/cad-reminders.js
@@ -140,14 +140,10 @@ describe('lib/cad-reminders', () => {
     describe('process', () => {
       let before, processResult;
 
-      beforeEach((done) => {
+      beforeEach(async () => {
         before = Date.now();
-        cadReminders.create('blee').then(() => {
-          setTimeout(async () => {
-            processResult = await cadReminders.process();
-            done();
-          }, 2);
-        });
+        await cadReminders.create('blee', before);
+        processResult = await cadReminders.process(before + 2);
       });
 
       afterEach(async () => {

--- a/packages/fxa-auth-server/test/local/verification-reminders.js
+++ b/packages/fxa-auth-server/test/local/verification-reminders.js
@@ -130,14 +130,15 @@ describe('lib/verification-reminders:', () => {
     describe('process:', () => {
       let before, processResult;
 
-      beforeEach((done) => {
+      beforeEach(async () => {
         before = Date.now();
-        verificationReminders.create('blee').then(() => {
-          setTimeout(async () => {
-            processResult = await verificationReminders.process();
-            done();
-          }, 2);
-        });
+        await verificationReminders.create(
+          'blee',
+          undefined,
+          undefined,
+          before
+        );
+        processResult = await verificationReminders.process(before + 2);
       });
 
       afterEach(() => {
@@ -242,9 +243,10 @@ describe('lib/verification-reminders:', () => {
   });
 
   describe('create with metadata:', () => {
-    let createResult;
+    let before, createResult;
 
     beforeEach(async () => {
+      before = Date.now();
       createResult = await verificationReminders.create('wibble', 'blee', 42);
     });
 
@@ -299,11 +301,8 @@ describe('lib/verification-reminders:', () => {
     describe('process:', () => {
       let processResult;
 
-      beforeEach((done) => {
-        setTimeout(async () => {
-          processResult = await verificationReminders.process();
-          done();
-        }, 2);
+      beforeEach(async () => {
+        processResult = await verificationReminders.process(before + 2);
       });
 
       it('returned the correct result', async () => {
@@ -392,11 +391,10 @@ describe('lib/verification-reminders:', () => {
       describe('process:', () => {
         let secondProcessResult;
 
-        beforeEach((done) => {
-          setTimeout(async () => {
-            secondProcessResult = await verificationReminders.process();
-            done();
-          }, 1000);
+        beforeEach(async () => {
+          secondProcessResult = await verificationReminders.process(
+            before + 1000
+          );
         });
 
         // NOTE: Because this suite has a slow setup, don't add any more test cases!

--- a/packages/fxa-auth-server/test/oauth/api.js
+++ b/packages/fxa-auth-server/test/oauth/api.js
@@ -188,7 +188,7 @@ describe('/v1', function () {
   const VERIFY_FAILURE = '{"status": "failure"}';
 
   before(async function () {
-    this.timeout(20000);
+    this.timeout(30000);
     Server = await testServer.start();
     return Promise.all([
       genAssertion(USERID + config.get('oauthServer.browserid.issuer')).then(

